### PR TITLE
search: check patterns for field prefix

### PR DESCRIPTION
**Pull request by rvantonder**
Part of #13128. Fixes a minor issue where prior to this PR:

`(repocommitafter:"thing with a space")` => interpreted as a search pattern `(repocommitafter:"thing with a space")`.

Desired meaning: interpret this as `repocommitafter:"thing with a space"`.

First commit performs the change and adds the test. The second commit factors out the shared code we use for parsing and checking, and updates some tests.

---

Long version:

The previous behavior erroneously interpreted a pattern instead of the desired group containing parameters when the group syntax `(...)` contained a parameter with a quoted value containing a space. This is because the search pattern parser heuristic would terminate at a space, and not realize that the quotes mean that it is a single value. Because its too tricky to try and also recognize well-quoted values for these in that function, I'm changing the heuristic to recognize valid fields by prefix earlier, rather than consider the values.

### CODE REVIEW

File: internal/search/query/literal_parser.go [hunk 1]
```diff
@@ -33,17 +33,17 @@ func ScanAnyPatternLiteral(buf []byte) (scanned string, count int) {
 	return scanned, count
 }
 
-// isParameter returns whether the token is a parameter.
-func isParameter(token []byte) bool {
-	parser := &parser{buf: token}
-	_, ok, _ := parser.ParseParameter()
-	return ok
+// isField returns whether the prefix of the string matches a recognized field.
```
Comment (stefanhengl) [inline internal/search/query/literal_parser.go:36]:
```suggestion
// isField returns whether the prefix of buf matches a recognized field.
```

File: internal/search/query/literal_parser.go [hunk 2]
```diff
@@ -33,17 +33,17 @@ func ScanAnyPatternLiteral(buf []byte) (scanned string, count int) {
...
+func isField(buf []byte) bool {
+	field, _, _ := ScanField(buf)
+	return field != ""
 }
 
 // ScanBalancedPatternLiteral attempts to scan parentheses as literal patterns.
 // It returns the scanned string, how much to advance, and whether it succeeded.
 // Basically it scans any literal string, including whitespace, but ensures that
-// a resulting string does not contain 'and' or 'or keywords, and is balanced.
+// a resulting string does not contain 'and' or 'or keywords, nor paramters, and
```
Comment (stefanhengl) [inline internal/search/query/literal_parser.go:45]:
```suggestion
// a resulting string does not contain 'and' or 'or keywords, nor parameters, and
```

**Review by stefanhengl — APPROVED** [review]
Looks good! Just nits.

File: internal/search/query/literal_parser.go (hunk 2 above)
Comment (rvantonder) [reply internal/search/query/literal_parser.go:45]: will address up the stack

File: internal/search/query/literal_parser.go (hunk 1 above)
Comment (rvantonder) [reply internal/search/query/literal_parser.go:36]: will address up the stack

**Pull request merged**